firebase-firestore-ktx/src/test/java/com/google/firebase/firestore/FirestoreTests.kt

```diff
@@ -0,0 +1,68 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore
+
+import com.google.common.truth.Truth.assertThat
+import com.google.firebase.Firebase
+import com.google.firebase.FirebaseApp
+import com.google.firebase.app
+import com.google.firebase.options
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+import org.robolectric.RuntimeEnvironment
+
+const val APP_ID = "APP_ID"
+const val API_KEY = "API_KEY"
+
+const val EXISTING_APP = "existing"
+
+@RunWith(RobolectricTestRunner::class)
+class FirestoreTests {
+    @Before
+    fun setup() {
+        FirebaseApp.initializeApp(
+            RuntimeEnvironment.application,
+            Firebase.options(applicationId = APP_ID, apiKey = API_KEY) {
+                projectId = "123"
+            }
+        )
+
+        FirebaseApp.initializeApp(
+            RuntimeEnvironment.application,
+            Firebase.options(applicationId = APP_ID, apiKey = API_KEY) {
+                projectId = "123"
+            },
+            EXISTING_APP
+        )
+    }
+
+    @After
+    fun cleanup() {
+        FirebaseApp.clearInstancesForTest()
+    }
+
+    @Test
+    fun `firestore should delegate to FirebaseFirestore#getInstance()`() {
+        assertThat(Firebase.firestore).isSameAs(FirebaseFirestore.getInstance())
+    }
+
+    @Test
+    fun `FirebaseApp#firestore should delegate to FirebaseFirestore#getInstance(FirebaseApp)`() {
+        assertThat(Firebase.app.firestore).isSameAs(FirebaseFirestore.getInstance(Firebase.app))
```

Comment:
fwiw this is the pattern we have today in our js sdks: e.g.
```javascript
firebase.app('myApp').firestore
```

Speaking of Swift, afaik the reason extension methods are not used there because we don't have first-class Swift SDKs yet, but we have our ObjC SDKs annotated in a way that makes them more friendly when used from Swift, for this reason creating extension from ObjC methods is not an option atm.

Also the pattern of having `app` as the first parameter in both Java and ObjC is the necessary mechanism as there is no way to extend classes from outside(with e.g. extension methods). But in the presence of extension methods imo it is logical to use them for this use case as they have a couple of advantages in terms of API terseness. e.g. compare the following:
```kotlin
// function taking app as first argument
Firebase.firestore(Firebase.app("foo"), "dbName")

// extension method
Firebase.app("foo").firestore("dbName)

// or even
StringUtils.plus("foo", "bar")
// vs
"foo".plus("bar")
```

Reply:
Let's keep discussing this issue, like I said this API is not final and we can iterate on it. But for now I'd like to submit it as is(note the branch name is not master, its `kotlin_support`) to keep making progress as it's hackweek after all, not a discussion/meetings week :)